kmod/core/core.c

```diff
@@ -341,27 +357,43 @@ static int kpatch_apply_patch(void *data)
 			hash_del_rcu(&func->node);
 		} while_for_each_linked_func();
 
-		return -EBUSY;
+		ret = -EBUSY;
+		goto err;
 	}
 
-	/* run any user-defined load hooks */
+	/* run any user-defined post-patch callbacks */
 	list_for_each_entry(object, &kpmod->objects, list) {
-		if (!kpatch_object_linked(object))
+		if (!kpatch_object_linked(object) ||
+		    !object->post_patch_callback)
 			continue;
-		list_for_each_entry(hook, &object->hooks_load, list)
-			(*hook->hook)();
-	}
 
+		(*object->post_patch_callback)(object);
+	}
 
 	return 0;
+err:
+	kpatch_state_finish(KPATCH_STATE_FAILURE);
+
+	WARN(ret, "error (%d) patching '%s'\n", ret,
+		object->name ? object->name : "vmlinux");
+
+	/* run any user-defined post-unpatch callbacks */
+	list_for_each_entry_continue_reverse(object, &kpmod->objects, list) {
+		if (!kpatch_object_linked(object) ||
+		    !object->post_unpatch_callback)
```

Comment:
This seems to require `!object->callbacks_enabled` check, otherwise we are calling post_unpatch_callback() even in the case when pre_patch_callback() fails which is not consistent with it's behavior in livepatch and kpatch's kpatch_remove_patch().

Reply:
If a pre-patch callback fails, the `object` iterator is re-used in this list traversal (ie, `_continue`, and `_reverse`), so it should only go back over the previous objects that did not fail.

I can explain as much in the "run any user-defined post-unpatch callbacks" comment, or just modify the list iteration to go over the entire list, skipping the non-enabled ones.

Reply:
Right, that's just me missing the "_continue" part. A comment explaining this might make it easier to spot.